include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -183,38 +183,83 @@ struct __init_processing
 //------------------------------------------------------------------------
 // transform_reduce
 //------------------------------------------------------------------------
-
-template <typename _ExecutionPolicy, typename _Operation1, typename _Operation2>
-struct transform_init
+template <typename _ExecutionPolicy, typename _Tp, typename _Operation1, typename _Operation2,
+          const ::std::size_t __iters_per_work_item = 0>
+struct transform_reduce
 {
+  private:
+    // Apply unary and binary operations to __items_to_process values and return result.
+    template <typename _BinaryOp, typename _UnaryOp, typename... _Acc>
+    static constexpr _Tp
+    apply_unary_binary(const ::std::size_t __items_to_process, const ::std::size_t __adjusted_global_id,
+                       _BinaryOp __binary_op, _UnaryOp __unary_op, const _Acc&... __acc)
+    {
+        // Keep these statements in the same scope to allow for better memory alignment.
+        // Vectorized access if __items_to_process is known at compile time.
+        _Tp __res = __unary_op(__adjusted_global_id, __acc...);
+        for (::std::size_t __i = 1; __i < __items_to_process; ++__i)
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + __i, __acc...));
+        return __res;
+    }
+
+  public:
     _Operation1 __binary_op;
     _Operation2 __unary_op;
 
-    template <typename _NDItemId, typename _Size, typename _AccLocal, typename... _Acc>
+    template <typename _Size, typename... _Acc>
+    _Tp
+    seq(const _Size __n, const _Acc&... __acc) const
+    {
+        return apply_unary_binary(__n, 0, __binary_op, __unary_op, __acc...);
+    }
+
+    // This routine will do nothing if the work item operates outside the bounds of __n. However, using checks
+    // (__global_idx * __iters_per_work_item) < __n) within the calling routine instead degrades the performance.
+    template <typename _Size, typename _AccLocal, typename... _Acc>
     void
-    operator()(const _NDItemId __item, _Size __n, ::std::size_t __iters_per_work_item, ::std::size_t __global_id,
-               ::std::size_t __global_offset, _AccLocal& __local_mem, const _Acc&... __acc) const
+    parallel(const ::std::uint16_t __local_id, const _Size __n, const ::std::size_t __dynamic_iters_per_work_item,
+             const ::std::size_t __global_id, const ::std::size_t __global_offset, _AccLocal& __local_mem,
+             const _Acc&... __acc) const
     {
-        ::std::size_t __adjusted_global_id = __global_offset + __iters_per_work_item * __global_id;
-        _Size __adjusted_n = __global_offset + __n;
-        if (__adjusted_global_id < __adjusted_n)
+        ::std::size_t __adjusted_global_id = __global_offset;
+        ::std::size_t __items_to_process = __n;
+        if constexpr (__iters_per_work_item > 0)
```

Comment:
I that case (many the similar "if") I tend to conclusion, that having  a specialization of `struct transform_reduce` for compile time case is better :)